### DISCUSSION TITLE
How to move camera poses in colmap?

### DISCUSSION THREAD

**Discussion by misza222pxo** [colmap/colmap#2741] [Q&A]
I have images.txt file with 2 records:

```
    1 0.01733147722376776 0.37417104773651677 0.042901493532558815 -0.9262046797472335 2.0 2.0 2.0 0 img0.jpeg

    2 -0.2 0.5 0.04 0.1 1.34681217 1.08815916 0.04479532 0 img0.jpeg
    
```
![initial cameras](https://github.com/user-attachments/assets/f1ddbbe6-69c1-4d2d-a1d1-c8dcd7b5d01f)

I try to create a new camera position centred where the camera 1 is, but it's direction vector is the same as camera 2 direction. 
My steps are:
1. get centre of the camera 1
   After [colmap docs](https://colmap.github.io/format.html#images-txt) the camera center is located at `-R^t * T`
   ```
   from scipy.spatial.transform import Rotation as R

   def compute_camera_center(quaternion, translation):
      rotation = R.from_quat(quaternion)
      R_matrix = rotation.as_matrix()
      camera_center = -R_matrix.T @ translation
      return camera_center

   compute_camera_center([0.01733147722376776,
                          0.37417104773651677,
                          0.042901493532558815,
                          -0.9262046797472335], 
                         [-2.68882953, -2.17631834, -0.18394084])
   # result array([ 1.90907645,  2.4155578 , -1.58761067])
   ```
2. get a translation vector for camera 2 quaternion and camera 1 center
   This is inverting previous equation
   ```
   from scipy.spatial.transform import Rotation as R

   def compute_new_translation(quaternion, camera_center):
      rotation = R.from_quat(quaternion)
      R_matrix = rotation.as_matrix()
      new_translation = -R_matrix @ camera_center
      return new_translation

   compute_new_translation([ 1.90907645,  2.4155578 , -1.58761067], [-0.2, 0.5, 0.04, 0.1])
   # result array([ 3.38416821, -0.11275296, -0.7312266 ])
   ```

So the final file with all 3 cameras looks like:
```
    1 0.01733147722376776 0.37417104773651677 0.042901493532558815 -0.9262046797472335 2.0 2.0 2.0 0 img0.jpeg

    2 -0.2 0.5 0.04 0.1 1.34681217 1.08815916 0.04479532 0 img0.jpeg
    
    3 -0.2 0.5 0.04 0.1 3.38416821 -0.11275296 -0.7312266 0 img1.png
    
```

But the third camera ends up being located somewhere else then where camera 1 is. 

![239sOJM6](https://github.com/user-attachments/assets/69a4f553-97af-4149-9f25-c995ec23f89a)

Does anyone knows how to compute translation verctor for the camera 3, so it has direction of camera 2, but is located where camera 1 is?

![oIQZ7sA4](https://github.com/user-attachments/assets/725c9cae-bea9-4669-b47a-a68bb6695fae)

**Comment by misza222pxo** [comment]
Here is the minimal dataset that I was working with, so you can open it in colmap
[colmap_data.zip](https://github.com/user-attachments/files/16819529/colmap_data.zip)

**Comment by Goulustis** [comment]
I believe you just need to make a copy of the camera2's camera matrix and overwrite the translation with the translation component from that of camera1

**Comment by misza222pxo** [comment]
It turned out, that all the math was correct, the only problem was that quaternions were not normalized before computing centers and translation

**Reply by mattiasmar** [reply]
Hello @misza222pxo  
How did your final code look like?
```
def foo(new_position, new_direction):
 ...
 return qvec, tvec
```

**Reply by mattiasmar** [reply]
I normalized the quaternions of image 2 & 3 and camera 1 & 3 are still not aligned in the colmap viewer:
![image](https://github.com/user-attachments/assets/3450d331-bd2c-4c2e-8d0a-eb8020fd4a13)
